### ISSUE TITLE
CUDA elementwise binary ops reject 0-D scalar broadcast in traced linalg AD

### ISSUE THREAD

**Issue by terasakisatoshi** [tensor4all/tenferro-rs#1354]
## Summary

CUDA/CubeCL elementwise binary ops currently reject shape pairs like `[N]` and `[]` before launching the GPU kernel. This makes mechanically generated traced linalg AD graphs fail on CUDA when they contain vector/scalar operations, even though the same graph is valid on CPU.

This surfaced while adding a GPU linalg JVP/VJP benchmark in `tenferro-benchmark`: `grad_sum_qr` JVP/VJP fails for `tenferro-cuda-trace` with a runtime shape error.

## Reproduction

In `tenferro-benchmark`, run the GPU linalg AD runner for a QR problem:

```bash
./target/release/benchmark_gpu_linalg_ad /tmp/qr_jvp.jsonl 0 \
  linalg_ad_small_grad_sum_qr_jvp_f64_2 -- benchmarks/gpu/linalg_jvp_vjp.yaml
```

Observed before the local experimental fix:

```text
runtime_failed: warmup: mul: shape mismatch lhs=[2] rhs=[]
```

The same failure pattern appears for larger QR JVP/VJP cases, with `lhs=[N] rhs=[]`.

## Expected behavior

CUDA backend binary elementwise ops should match CPU backend semantics for 0-D scalar operands:

- `tensor_shape * []` should broadcast the scalar to `tensor_shape`
- `[] * tensor_shape` should broadcast the scalar to `tensor_shape`

This should apply at least to same-dtype elementwise ops routed through the CubeCL binary dispatch path (`add`, `sub`, `mul`, etc.).

## Actual behavior

`crates/tenferro-gpu/src/cubecl/dispatch.rs` routes binary elementwise ops through `launch_binary`, which validates raw binary shapes as equal before launching. A 0-D scalar paired with a vector/matrix therefore fails before reaching a kernel:

```text
mul: shape mismatch lhs=[N] rhs=[]
```

## Impact

This is a backend semantic gap rather than missing QR AD support:

- CUDA QR primal works.
- CPU traced QR JVP/VJP works.
- The generated QR AD graph is valid but hits CUDA binary scalar broadcast semantics.
- GPU linalg AD benchmark currently has to mark `grad_sum_qr` JVP/VJP as unsupported for `tenferro-cuda-trace`.

## Potential fix

In the CubeCL binary elementwise dispatch path, detect 0-D scalar operands and broadcast the scalar on GPU to the other operand's shape before calling the existing same-shape binary kernel. A local smoke test with this approach made both of these pass:

```text
linalg_ad_small_grad_sum_qr_jvp_f64_2 ok
linalg_ad_small_grad_sum_qr_vjp_f64_2 ok
```

A source-level regression test should also pin that `launch_binary_elementwise_kernel` keeps explicit scalar-lhs and scalar-rhs broadcast branches.

**Comment by terasakisatoshi** [comment]
## Self-contained benchmark context

The failing benchmark is intended to compare CUDA linalg AD for the scalar loss

```text
loss(A) = sum(Q) + sum(R), where (Q, R) = qr(A)
```

and then measure both:

```text
JVP: d loss(A)[T]
VJP: d loss(A)^T [1.0]
```

The input matrix is a deterministic well-conditioned `float64` matrix. A simplified version of the data generator is:

```python
import numpy as np

def data(shape, seed):
    size = int(np.prod(shape))
    values = []
    for i in range(size):
        x = (i * 6364136223846793005 + seed * 1442695040888963407) & ((1 << 64) - 1)
        values.append(((x % 1024) - 512) / 512.0)
    return np.array(values, dtype=np.float64).reshape(shape)

def well_conditioned(n, seed):
    matrix = data((n, n), seed) * 0.05
    for j in range(n):
        matrix[j, j] += 1.0 + j / max(n, 1)
    return matrix

def tangent_seed(matrix_seed):
    return matrix_seed + 1000
```

The PyTorch CUDA reference used by the benchmark is essentially:

```python
import time
import statistics
import torch
from torch.func import jvp, vjp


def torch_loss_qr(a):
    q, r = torch.linalg.qr(a)
    return q.sum() + r.sum()


def bench(fn, sync, runs=7, warmups=3):
    for _ in range(warmups):
        sync(fn())
    times = []
    for _ in range(runs):
        start = time.perf_counter()
        sync(fn())
        times.append((time.perf_counter() - start) * 1000.0)
    values = sorted(times)
    return statistics.median(values), values[(3 * len(values)) // 4] - values[len(values) // 4]


def run_torch_qr_jvp(n, matrix_seed, device="cuda:0"):
    matrix = well_conditioned(n, matrix_seed)
    tangent = data((n, n), tangent_seed(matrix_seed))
    x = torch.tensor(matrix, dtype=torch.float64, device=device)
    t = torch.tensor(tangent, dtype=torch.float64, device=device)

    def runner():
        return jvp(torch_loss_qr, (x,), (t,))[1]

    return bench(runner, lambda value: torch.cuda.synchronize(device))


def run_torch_qr_vjp(n, matrix_seed, device="cuda:0"):
    matrix = well_conditioned(n, matrix_seed)
    x = torch.tensor(matrix, dtype=torch.float64, device=device)
    cotangent = torch.tensor(1.0, dtype=torch.float64, device=device)

    # Important for fair timing: build the VJP closure inside the timed region.
    # Otherwise the QR forward pass is measured outside the timed region.
    def runner():
        _, vjp_fn = vjp(torch_loss_qr, x)
        return vjp_fn(cotangent)[0]

    return bench(runner, lambda value: torch.cuda.synchronize(device))
```

On the tenferro side, the benchmark builds an equivalent traced CUDA graph:

```rust
let a = upload_traced(rt, &[n, n], well_conditioned_matrix(n, matrix_seed))?;
let (q, r) = a.qr()?;
let q_sum = q.reduce_sum(&[0, 1])?;
let r_sum = r.reduce_sum(&[0, 1])?;
let loss = (&q_sum + &r_sum)?;

// JVP
let tangent = upload_traced(rt, &[n, n], data_for_shape(&[n, n], tangent_seed))?;
let out = ad_context().jvp(&loss, &a, &tangent)?;

// VJP
let cotangent = upload_traced(rt, &[], vec![1.0])?;
let out = ad_context().vjp(&loss, &a, &cotangent)?;
```

The traced output is compiled once and each timed run executes:

```rust
let t0 = Instant::now();
let out = executor.run_many(&program)?;
executor.backend().runtime().synchronize()?;
times_ms.push(t0.elapsed().as_secs_f64() * 1000.0);
```

The failing command in the benchmark repository was:

```bash
./target/release/benchmark_gpu_linalg_ad /tmp/qr_jvp.jsonl 0 \
  linalg_ad_small_grad_sum_qr_jvp_f64_2 -- benchmarks/gpu/linalg_jvp_vjp.yaml
```

Before the local experimental scalar-broadcast fix, the tenferro CUDA trace path produced:

```text
runtime_failed: warmup: mul: shape mismatch lhs=[2] rhs=[]
```

The same issue occurs for QR VJP and larger `n`, e.g. `lhs=[256] rhs=[]`. PyTorch CUDA completes the equivalent JVP/VJP benchmark, and the tenferro CPU backend has CPU-style scalar handling, so this points to the CUDA/CubeCL binary elementwise dispatch semantics rather than to QR AD being absent.
